### ISSUE TITLE
ARM64 Support

### ISSUE THREAD

**Issue by faddat** [pikvm/pi-builder#4]
Some of our software has known [security issues in 32 bit environments](https://github.com/tendermint/tendermint/blob/master/docs/tendermint-core/running-in-production.md#validator-signing-on-32-bit-architectures-or-arm).

Would you accept a patch that extends pi-builder to support the creation of 64 bit images?

**Comment by mdevaev** [comment]
Hi! Yes, why not. Check the commit #6d5ccc76a2bc7b0b829a94ec99150accac53e3df to start. I used to have aarch64 support, but I removed it because I didn't use it. it seems that there was a basic working capacity, but it was a long time ago and I don't remember the details anymore. Note that QEMU 5.1 is [broken](https://bugs.launchpad.net/qemu/+bug/1891748), so you need to use QEMU 5.0.1 right now.

**Comment by faddat** [comment]
That's wonderful.

I haven't made the transition yet but my expectation is that I will be able to use your software to dramatically reduce our image build times.

**Comment by mdevaev** [comment]
Ha-ha, well, that's the main reason I did it :) This helped me a lot when developing Pi-KVM, otherwise I could have waited hours for the build to complete.

**Comment by faddat** [comment]
So even with us, build times are still kind of an issue for me. I am hoping to figure out a way to make this even faster and possibly slim down arch linux arm.  `bout 5 years ago I remember setting up Arch for an Orange Pi, and I recall just using pacstrap.   I don't know if it's still possible or what might be making the initial download the 400 MB that it is but I am hoping to reduce that as well.

For the kind of thing I'm working on, less software ends up being exponentially more.

**Comment by faddat** [comment]
If it's not currently desired then I can tell you that the default raspberry Pi 64-bit Colonel is actually not all that stable and you probably don't want to use it so I think that for now I'm going to close this issue.

I solved the problem by using a user contributed kernel that more closely matches what the raspberry Pi inc developers were doing.

**Comment by mdevaev** [comment]
Okay, If you want to go back to this issue, I won't mind.

**Comment by faddat** [comment]
OK, I'll leave it open-- when mainline smoothly supports Pi 4, it'll make good sense to add arm64 support, I reckon.

**Comment by mdef** [comment]
I'm also interested to have 64bit build 
or can you confirm if it possible to run 32bit packages what you provide on aarch64 arch  linux?

**Comment by mdevaev** [comment]
I haven't tested this case, so I can't answer.

**Comment by mdef** [comment]
I'm actually interested to have 64bit arch linux to run pikvm on it, and other sortware in docker.
Will it work this way:
- install ArchLinuxARM-rpi-aarch64-latest.tar.gz
- install precompiled pikvm for aarch64 if the exists somewhere
Thank y

**Comment by mdevaev** [comment]
64bit is not compatible with tc358743. perhaps something else is not working, I have not studied this issue in detail

**Comment by faddat** [comment]
> I'm actually interested to have 64bit arch linux to run pikvm on it, and other sortware in docker.
> Will it work this way:
> 
> * install ArchLinuxARM-rpi-aarch64-latest.tar.gz
> * install precompiled pikvm for aarch64 if the exists somewhere
>   Thank you in advance

No, but it will work this way:

https://github.com/faddat/sos/blob/dcdfa6e029d683d38e379389f2fdad7f393bd0c0/Dockerfile#L29-L33

Those lines are in fact super important. Also, if you'd like, you can just try forking SOS which is more or less a derivative of pi-builder here in fact.  The biggest difference between SOS and Pye builders is probably that I offer the user fewer options, and don't really do very much parameterization or use of variables. The build script is quite linear.

**Comment by faddat** [comment]
So the thing that I'd like to contribute I think is actually the use of Docker buildx.

I'm also going to look over that PR that adds multi-device support because I'm tinkering with the same stuff in SOS.

**Comment by Zitt** [comment]
Coming up on two years later and no progress :(

**Comment by mdevaev** [comment]
Hey guys, better late than never, right?

I have added basic aarch64 support. Not all stages work with them (there is no watchdog, for example), but in general it should work.

However, I caution against further use of pi-builder on x86_64. It will be better to use it on arm or aarch64, i.e. build Raspberry on Raspberry. The reason is that QEMU can fail unpredictably inside different installation scripts in the package manager, and this can lead to unpredictable behavior and generation of faulty OS images. The new QEMU works poorly, QEMU for AArch64 is even more unstable, and the only thing that works well now is old QEMU armv7 on x86_64. I have abandoned cross-assembly and use a cluster of several Raspberries for my purposes.

**Comment by onedr0p** [comment]
Wow that's awesome, can't wait download the image and flash it when it's all ready

**Comment by onedr0p** [comment]
@mdevaev is there an issue for tracking that will support building/publishing arm64 images for PiKVM? I would like to use the PiKVM for a few more things but the armv7 arch is limiting.

**Comment by mdevaev** [comment]
I'm not working on it right now, but I have some plans for this year.

**Comment by kaineub** [comment]
@mdevaev is there any update on that? I also find the arch pretty limiting right now.

**Comment by mdevaev** [comment]
Reopening since I'm working on 64 bit PiKVM.

**Comment by mdevaev** [comment]
While I'm working on this, can someone tell me about the use case? Why do you need 64 bit, and why didn't 32 bit meet your needs?

**Comment by onedr0p** [comment]
@mdevaev since the pikvm is a device that is running all the time I'd like to be able to run additional tools on it like bind, blocky, matchbox, docker etc... the only problem with this currently is the limited support for the armv7 architecture projects either take a long time to compile, or won't at all or docker containers do not exist.

While I understand configuring the pikvm to do more than it's expected to is not supported the tools I would run wouldn't shorten the life of the sd card nor interfere with pikvm software.

**Comment by mdevaev** [comment]
@onedr0p Got it, thank you for the info.

Anybody else? :)

**Comment by kaineub** [comment]
I have a recovery tool that relies on having a 64-bit python interpreter. I currently need to have a second machine hooked up in addition to the PiKVM to use it.

**Comment by mdevaev** [comment]
I've built some images for PiKVMs based on Pi 4:
* [v4plus-hdmi-rpi4](https://files.pikvm.org/images/v4plus-hdmi-rpi4/aarch64/v4plus-hdmi-rpi4-aarch64-latest.img.xz)
* [v4mini-hdmi-rpi4](https://files.pikvm.org/images/v4mini-hdmi-rpi4/aarch64/v4mini-hdmi-rpi4-aarch64-latest.img.xz)
* [v3-hdmi-rpi4](https://files.pikvm.org/images/v3-hdmi-rpi4/aarch64/v3-hdmi-rpi4-aarch64-latest.img.xz)
* [v3-hdmi-rpi4-box](https://files.pikvm.org/images/v3-hdmi-rpi4/aarch64/v3-hdmi-rpi4-aarch64-box-latest.img.xz)
* [v2-hdmi-rpi4](https://files.pikvm.org/images/v2-hdmi-rpi4/aarch64/v2-hdmi-rpi4-aarch64-latest.img.xz)
* [v2-hdmiusb-rpi4](https://files.pikvm.org/images/v2-hdmiusb-rpi4/aarch64/v2-hdmiusb-rpi4-aarch64-latest.img.xz)

**Comment by onedr0p** [comment]
@mdevaev moving forward are updates also supported on this arch? Meaning pikvm-update will pull arm64?

**Comment by mdevaev** [comment]
You need to reinstall it, update will not give you a new arch.

**Comment by onedr0p** [comment]
Yeah that I understand I was wondering if updates after reinstalling are supported right now.

**Comment by mdevaev** [comment]
I see 3 likes, anybody tested? :)

**Comment by onedr0p** [comment]
I should have time within a couple days to reinstall and migrate.

**Comment by mdevaev** [comment]
👌

**Comment by onedr0p** [comment]
Nice work @mdevaev I was able to find time to reinstall and haven't discovered any issues yet.

```
uname -a
Linux pikvm 6.6.45-10-rpi #1 SMP PREEMPT Sun Feb 16 22:50:37 UTC 2025 aarch64 GNU/Linux
```

**Comment by buroa** [comment]
Works great here too, thanks!

**Comment by szinn** [comment]
Does the image above tagged with v4+ work on a prebuilt v4+ with the cm4 RPI?

**Comment by mdevaev** [comment]
v4 image is for our own V4 Plus branded devices

**Comment by mdevaev** [comment]
Okay, I've got some confirmations that it's working, and tested myself too. Now all images will be 64 bit except RPi2. Thank you for testing.

**Comment by mihcox** [comment]
what is the difference between these 2 images?

[v3-hdmi-rpi4](https://files.pikvm.org/images/v3-hdmi-rpi4/aarch64/v3-hdmi-rpi4-aarch64-latest.img.xz)
[v3-hdmi-rpi4-box](https://files.pikvm.org/images/v3-hdmi-rpi4/aarch64/v3-hdmi-rpi4-aarch64-box-latest.img.xz)

**Comment by onedr0p** [comment]
By looking at the links on the [download](https://pikvm.org/download/) page for the 32-bit version it looks like...

> [v3-hdmi-rpi4](https://files.pikvm.org/images/v3-hdmi-rpi4/aarch64/v3-hdmi-rpi4-aarch64-latest.img.xz)

is the DIY hat kit

> [v3-hdmi-rpi4-box](https://files.pikvm.org/images/v3-hdmi-rpi4/aarch64/v3-hdmi-rpi4-aarch64-box-latest.img.xz)

is the pre-assembled pikvm
